File: mobile/lib/features/games/domain/entities/task_scheduling_metrics.dart
import 'game_metrics.dart';

/// Métriques objectives du « Planning journalier » (Planifik #2).
///
/// Mesures collectées pendant la partie — jamais un score. Le serveur (ou le
/// mock hors ligne) applique le barème /10. Aligné sur `TaskSchedulingMetrics`
/// du contrat `games.openapi.yaml`.
///
/// Des compteurs, et non des booléens : les trois premières composantes du
/// barème client se calculent sur des RATIOS. L'ancien contrat tout-ou-rien
/// donnait le même zéro à qui tenait huit contraintes sur neuf et à qui n'en
/// tenait aucune.
class TaskSchedulingMetrics extends GameMetrics {
  const TaskSchedulingMetrics({
    required this.universeId,
    required this.dependencyEdgeCount,
    required this.dependencyEdgesRespected,
    required this.directDependencyViolations,
    required this.timingConstraintCount,
    required this.timingConstraintsRespected,
    required this.collisionFree,
    required this.deadTimeRatio,
    required this.proactiveAdjustments,
    required this.reactiveAdjustments,
    this.levelsPlayed = 1,
    this.planningLatencyMs,
  });

  /// Univers tiré, pour la traçabilité d'une passation à l'autre.
  final String universeId;

  final int dependencyEdgeCount;
  final int dependencyEdgesRespected;

  /// Tâches lancées sans que leur prérequis soit terminé.
  ///
  /// Pénalisées à part : le référentiel y voit une erreur de règle, pas un
  /// simple sous-optimum d'ordre.
  final int directDependencyViolations;

  /// n RÉEL de contraintes horaires de l'univers — 5 à 7 selon l'univers.
  final int timingConstraintCount;
  final int timingConstraintsRespected;

  /// Aucun bloc horaire fixe empêché de démarrer à son heure.
  final bool collisionFree;

  /// Part de temps mort sur l'amplitude du planning, dans [0,1].
  final double deadTimeRatio;

  /// Corrections décidées avant tout signal d'erreur.
  final int proactiveAdjustments;

  /// Corrections déclenchées après un signal d'erreur.
  final int reactiveAdjustments;

  /// Nombre de plannings joués dans la partie.
  ///
  /// Les seuils d'autorégulation du référentiel — 1 puis 3 corrections — ont
  /// été calibrés sur UN planning. Une partie en enchaîne plusieurs, et les
  /// compteurs ci-dessus en font la somme : sans ce nombre, le serveur
  /// appliquerait un barème de manche à un total de partie, et toute partie où
  /// le joueur se reprend une fois par manche tomberait à zéro.
  final int levelsPlayed;

  /// Temps de réflexion avant le premier placement.
  ///
  /// Métrique DIAGNOSTIQUE : le référentiel la veut au profil qualitatif et
  /// explicitement hors du score — elle distingue un profil impulsif d'un
  /// profil délibératif à score égal. `null` si aucune tâche n'a été placée.
  final int? planningLatencyMs;

  int get adjustmentCount => proactiveAdjustments + reactiveAdjustments;

  @override
  Map<String, dynamic> toJson() => {
    'universeId': universeId,
    'dependencyEdgeCount': dependencyEdgeCount,
    'dependencyEdgesRespected': dependencyEdgesRespected,
    'directDependencyViolations': directDependencyViolations,
    'timingConstraintCount': timingConstraintCount,
    'timingConstraintsRespected': timingConstraintsRespected,
    'collisionFree': collisionFree,
    'deadTimeRatio': deadTimeRatio,
    'proactiveAdjustments': proactiveAdjustments,
    'reactiveAdjustments': reactiveAdjustments,
    'levelsPlayed': levelsPlayed,
    if (planningLatencyMs != null) 'planningLatencyMs': planningLatencyMs,
  };
}

File: backend/src/main/java/com/zennyt/games/domain/vo/TaskSchedulingMetrics.java
package com.zennyt.games.domain.vo;

/**
 * Métriques brutes du mini-jeu « Planning journalier » (Planifik #2).
 *
 * <p>Mesures objectives collectées côté client — jamais un score. Le calcul
 * déterministe appartient au domaine ({@code PlanifikScoringService}), de sorte
 * que le client ne puisse pas s'auto-attribuer de points.
 *
 * <p>Les trois premières composantes du barème client se calculent sur des
 * RATIOS, d'où des compteurs plutôt que des booléens : un joueur qui respecte
 * huit contraintes horaires sur neuf ne vaut pas celui qui n'en respecte
 * aucune, et l'ancien contrat tout-ou-rien les confondait.
 *
 * @param universeId                identifiant de l'univers tiré, pour la traçabilité
 * @param dependencyEdgeCount       nombre total d'arêtes de dépendance de l'univers
 * @param dependencyEdgesRespected  arêtes effectivement respectées
 * @param directDependencyViolations tâches lancées sans que leur prérequis soit terminé
 * @param timingConstraintCount     n RÉEL de contraintes horaires de l'univers (5 à 7)
 * @param timingConstraintsRespected contraintes horaires tenues
 * @param collisionFree             aucun bloc horaire fixe empêché de démarrer à l'heure
 * @param deadTimeRatio             part de temps mort sur l'amplitude du planning [0,1]
 * @param proactiveAdjustments      corrections décidées avant tout signal d'erreur
 * @param reactiveAdjustments       corrections déclenchées par une erreur signalée
 * @param levelsPlayed              nombre de plannings joués — les seuils
 *                                  d'autorégulation valent PAR planning
 * @param planningLatencyMs         temps de réflexion avant le premier placement —
 *                                  métrique DIAGNOSTIQUE, explicitement hors du score
 */
public record TaskSchedulingMetrics(
    String universeId,
    int dependencyEdgeCount,
    int dependencyEdgesRespected,
    int directDependencyViolations,
    int timingConstraintCount,
    int timingConstraintsRespected,
    boolean collisionFree,
    double deadTimeRatio,
    int proactiveAdjustments,
    int reactiveAdjustments,
    int levelsPlayed,
    Integer planningLatencyMs
) implements GameMetrics {

    public TaskSchedulingMetrics {
        // Un client antérieur aux manches multiples n'envoie rien : on retombe
        // sur un planning, c'est-à-dire exactement l'ancien barème.
        if (levelsPlayed <= 0) {
            levelsPlayed = 1;
        }
        if (dependencyEdgeCount < 0 || dependencyEdgesRespected < 0
            || dependencyEdgesRespected > dependencyEdgeCount) {
            throw new IllegalArgumentException(
                "arêtes de dépendance incohérentes : "
                    + dependencyEdgesRespected + "/" + dependencyEdgeCount);
        }
        if (directDependencyViolations < 0) {
            throw new IllegalArgumentException("directDependencyViolations doit être >= 0");
        }
        if (timingConstraintCount < 0 || timingConstraintsRespected < 0
            || timingConstraintsRespected > timingConstraintCount) {
            throw new IllegalArgumentException(
                "contraintes horaires incohérentes : "
                    + timingConstraintsRespected + "/" + timingConstraintCount);
        }
        if (deadTimeRatio < 0.0 || deadTimeRatio > 1.0) {
            throw new IllegalArgumentException("deadTimeRatio hors [0,1] : " + deadTimeRatio);
        }
        if (proactiveAdjustments < 0 || reactiveAdjustments < 0) {
            throw new IllegalArgumentException("compteurs de corrections négatifs");
        }
        if (planningLatencyMs != null && planningLatencyMs < 0) {
            throw new IllegalArgumentException("planningLatencyMs doit être >= 0");
        }
    }

    /** Total des corrections, toutes natures confondues. */
    public int adjustmentCount() {
        return proactiveAdjustments + reactiveAdjustments;
    }
}
